libcloud/compute/drivers/vsphere.py

```diff
@@ -12,543 +12,216 @@
-        if ex_remove_files:
-            request = VI.Destroy_TaskRequestMsg()
-
-            _this = request.new__this(vm._mor)
-            _this.set_attribute_type(vm._mor.get_attribute_type())
-            request.set_element__this(_this)
-            ret = server._proxy.Destroy_Task(request)._returnval
-            task = VITask(ret, server)
-
-            # Wait for the task to finish
-            status = task.wait_for_state([task.STATE_SUCCESS,
-                                          task.STATE_ERROR])
-
-            if status == task.STATE_ERROR:
-                raise LibcloudError('Error destroying node: %s' %
-                                    (task.get_error_message()))
-        else:
-            request = VI.UnregisterVMRequestMsg()
+        try:
+            self.conn = connect.SmartConnect(host=host,
+                                             user=username,
+                                             pwd=password,
+                                             sslContext=context)
+            atexit.register(connect.Disconnect, self.conn)
+        except Exception as e:
+            if isinstance(e, vim.fault.InvalidLogin):
+                raise InvalidCredsError("username/password not acepted")
+            raise LibcloudError(str(e), driver=self)
+
+    @property
+    def _root(self):
+        return self.conn.RetrieveContent().rootFolder
+
+    @property
+    def _datacenters(self):
+        return [dc for dc in self._root.childEntity
+                if isinstance(dc, vim.Datacenter)]
+
+    @property
+    def _vms(self):
+        vms = []
+        nodes = (node for datacenter in self._datacenters
+                 for node in datacenter.vmFolder.childEntity)
+
+        def vms_in(folder):
+            """ recursively extract the vms from a folder """
+            for node in folder.childEntity:
+                if isinstance(node, vim.Folder):
+                    yield from vms_in(node)
+                elif isinstance(node, vim.VirtualMachine):
+                    yield node
+
+        for node in nodes:
+            if isinstance(node, vim.Folder):
+                vms.extend(vms_in(node))
+            elif isinstance(node, vim.VirtualMachine):
+                vms.append(node)
+
+        return vms
+
+    def _get_obj_by_name(self, vimtype, name):
+        if not isinstance(vimtype, list):
+            vimtype = [vimtype]
+
+        content = self.conn.RetrieveContent()
+        container = content.viewManager.CreateContainerView(
+            content.rootFolder,
+            vimtype,
+            True)
 
-            _this = request.new__this(vm._mor)
-            _this.set_attribute_type(vm._mor.get_attribute_type())
-            request.set_element__this(_this)
-            ret = server._proxy.UnregisterVM(request)
-            task = VITask(ret, server)
+        try:
+            return next(i for i in container.view if i.name == name)
+        except StopIteration:
+            return None
```

Comment:
You can give an additional argument to `next` to be a default value:

```python
return next((i for i in container.view if i.name == name), None)
```

But I don't care too much about that. This looks clean and explicit too.